Dashboard: fix tests and fixtures

jraddaoui opened the pull request:
Dashboard: fix administration and fpr tests

- Avoid database connection on `fpr.forms` module import. The connection
  was created to get the choices of some fields. However those choices
  are overwritten later in the forms constructor.
- Split, rename and move tests modules to be picked by `pytest`.
- Add `__init__` module to test packages to be picked by Django's runner.
- Change pytest ignore rules in tox.ini.

Dashboard: improve fixtures configuration

- Specify FIXTURE_DIRS setting using system agnostic paths.
- Do not include fixtures from archivematicaCommon.
- Import local instead of base settings in test settings to also include
  the debug changes and the fixture directories.
- Trust FIXTURE_DIRS setting in tests, which only makes necessary to
  include the filenames in the `fixtures` attribute.

Refs https://github.com/archivematica/Issues/issues/783.

jraddaoui commented:
Hi @ross-spencer and @sevein, this is ready for code review.

I decided to move the tests to the main `dashboard/tests` folder to avoid changing the [ignore rule](https://github.com/artefactual/archivematica/blob/qa/1.x/tox.ini#L27) from `tox.ini`. However, it may be better to keep the tests in `dashboard/src/fpr/tests/test_*.py` and `dashboard/src/components/administration/tests/test_*.py` to give them a little more context and remove the final `.../tests` from those ignore rules.

I also wonder, why do we have the MCPServer tests running separately? It's probably faster and easier to read to have them all in one block, since we're installing the same requirements in both.

Please, let me know your thoughts. Thanks!

ross-spencer reviewed (approved):
@jraddaoui we chatted in Slack too and I was able to see how this (didn't work) previously. And I can see the difference in the test count in the Dashboard tests now too:
```
86 passed, 6 skipped in 20.91 seconds
102 passed, 6 skipped in 32.09 seconds
```
Code looks great, and there's a lot here i'll come back to to learn from in future tests.

All in all, LGTM! :sunglasses: 

You also asked, about where these tests should sit. I think this is perfect. Knowing components like the FPR sit in the dashboard means looking in the dashboard test suite feels intuitive. I appreciate the idea of providing context. I feel like the balance is right now though.

MCP Server tests in .travis.yml separately? I see this. It feels like an oversight. Would it be worth setting up a development branch to pull the MCP server tests together with the others? I wonder if there's a legacy reason for them being separate that might reveal itself to need a little work? 

We could start with a new GitHub issue to capture that? What do you think?

sevein commented:
> MCP Server tests in .travis.yml separately? I see this. It feels like an oversight. Would it be worth setting up a development branch to pull the MCP server tests together with the others? I wonder if there's a legacy reason for them being separate that might reveal itself to need a little work?

I was trying to remember this the other day unsuccessfully. But I think I'm starting to remember now. We didn't have tests at all in MCPServer (https://github.com/archivematica/Issues/issues/88) and I wanted to write some first tests in preparation for the workflow work that I was going to be doing.
It wasn't only a pytest discovery issue but also a problem with Python module imports side-effects. The fix I brought was here: https://github.com/artefactual/archivematica/pull/1250.
IIRC my intention was to have a completely independent build in order to make that distinction with other components clearer. I wanted to have MCPServer using its own test settings (and added `src/MCPServer/lib/settings/test.py` for that purpose), making sure that the `settings.test` import was not resulting in loading the module of other components. And have that tested alone with Py2 and Py3.

> It's probably faster and easier to read to have them all in one block, since we're installing the same requirements in both.

We're not exactly installing the same requirements though, `testenv:mcpserver` purposely avoids installing `src/MCPClient/requirements/test.txt`. I'm actually not a big fan of having to maintain three apps (I'd rather have a Python package standard layout with multiple entry points) but my thought was to reflect closely how distributions have been deploying Archivematica (e.g. packages/ansible deploy independent venvs) and make sure that we do that early in our pipeline.

FWIW Travis seems to be running the builds in parallel. E.g. in build [555748329](https://travis-ci.org/artefactual/archivematica/builds/555748329), total time was 8 min 50 sec but wall-clock elapsed time was only 2 min 54 sec.

One question I have is whether we really want to move our tests around so they're discovered by pytest using the default test discovery settings, or on the contrary we want to tweak [pytest discoverry settings](https://docs.pytest.org/en/latest/example/pythoncollection.html) so it finds our tests where they're today.

> See https://docs.pytest.org/en/latest/reference.html#confval-python_files:
> By default, files matching test_*.py and *_test.py will be considered test modules.

I mentioned this in the original issue, see https://github.com/artefactual-labs/am/issues/29#issuecomment-426831225. To me the latter is preferably because it's already a common practice in Django to have tests inside the apps. It's also probably a scheme that scales better, since the intention of apps is to segregate responsibilities, as opposed to having a single directory for all our tests? There's also the Django test runner, which probably has its own app-aware discovery settings. I wonder if that's working at all or if we should pay attention to it.

jraddaoui commented:
> > MCP Server tests in .travis.yml separately? I see this. It feels like an oversight. Would it be worth setting up a development branch to pull the MCP server tests together with the others? I wonder if there's a legacy reason for them being separate that might reveal itself to need a little work?
> 
> I was trying to remember this the other day unsuccessfully. But I think I'm starting to remember now. We didn't have tests at all in MCPServer ([archivematica/Issues#88](https://github.com/archivematica/Issues/issues/88)) and I wanted to write some first tests in preparation for the workflow work that I was going to be doing.
> It wasn't only a pytest discovery issue but also a problem with Python module imports side-effects. The fix I brought was here: #1250.
> IIRC my intention was to have a completely independent build in order to make that distinction with other components clearer. I wanted to have MCPServer using its own test settings (and added `src/MCPServer/lib/settings/test.py` for that purpose), making sure that the `settings.test` import was not resulting in loading the module of other components. And have that tested alone with Py2 and Py3.
> 
> > It's probably faster and easier to read to have them all in one block, since we're installing the same requirements in both.
> 
> We're not exactly installing the same requirements though, `testenv:mcpserver` purposely avoids installing `src/MCPClient/requirements/test.txt`. I'm actually not a big fan of having to maintain three apps (I'd rather have a Python package standard layout with multiple entry points) but my thought was to reflect closely how distributions have been deploying Archivematica (e.g. packages/ansible deploy independent venvs) and make sure that we do that early in our pipeline.
> 
> FWIW Travis seems to be running the builds in parallel. E.g. in build [555748329](https://travis-ci.org/artefactual/archivematica/builds/555748329), total time was 8 min 50 sec but wall-clock elapsed time was only 2 min 54 sec.

Sounds good @sevein, in that case we may want to separate even more in the future, having really one block for each component. Nevertheless, if we go for the unification of the apps some day, this could be a good first step to make sure all the requirements are intalled and the test pass on the same env.

> One question I have is whether we really want to move our tests around so they're discovered by pytest using the default test discovery settings, or on the contrary we want to tweak [pytest discoverry settings](https://docs.pytest.org/en/latest/example/pythoncollection.html) so it finds our tests where they're today.
> 
> > See https://docs.pytest.org/en/latest/reference.html#confval-python_files:
> > By default, files matching test_*.py and *_test.py will be considered test modules.
> 
> I mentioned this in the original issue, see [artefactual-labs/am#29 (comment)](https://github.com/artefactual-labs/am/issues/29#issuecomment-426831225). To me the latter is preferably because it's already a common practice in Django to have tests inside the apps. It's also probably a scheme that scales better, since the intention of apps is to segregate responsibilities, as opposed to having a single directory for all our tests? There's also the Django test runner, which probably has its own app-aware discovery settings. I wonder if that's working at all or if we should pay attention to it.

Yeah, I personally prefer to have them in their related apps, however I'd try to avoid calling the modules `tests.py` and "force" the creation of a `tests` folder with the `test_*.py` modules. We'll only need to modify a little the ignore rules from `tox.ini`. Would you prefer that too?

sevein commented:
> I'd try to avoid calling the modules tests.py and "force" the creation of a tests folder with the test_*.py modules. We'll only need to modify a little the ignore rules from tox.ini. Would you prefer that too?

That would work for me. If you have a chance, please also test `manage.py test`. If that's a valid workflow (I'm still wondering if we should pay attention to it), we could subclass the discovery runner with a custom pattern - seehttps://docs.djangoproject.com/en/1.8/topics/testing/advanced/#defining-a-test-runner.

jraddaoui commented:
Thanks @sevein,

I've moved the tests and changed the `pytest` ignore rules. By adding the `__init__` modules to the test packages inside each app, they are recognized by Django's test runner. See [the current output](https://gist.github.com/jraddaoui/7cec90d023a5ea6e3d2010b4bae051b9) running `manage.py test` (sorry for the debug messages).

As you suspected, some changes will be needed to run the tests in `src/dashboard/tests` with Django's runner. We could tweak the discovery to look for tests in `src/dashboard` instead of in `src/dashboard/src` (the default top-level where `manage.py` is located). Or move the tests folder from `src/dashboard/tests` to `src/dashboard/src/tests` so they are found by default.

I tried the latter (+ adding the `__init__` module in the `tests` folder) and the results are not bad at all ([see output](https://gist.github.com/jraddaoui/608e910c29485de87269b8e5cab54d5f), again with a lot of debug, sorry). Nevertheless, it shows 5 failures from 81 tests, when the `pytest` run shows no failures from 102 tests, so this will require a little more effort than just the discovery issue and maybe it deserves its own issue.

ross-spencer commented:
Nice work Radda.

I can do a quick final review on this @sevein? But if it makes sense, I can
leave that for you? Let me know!

On Thursday, July 11, 2019, José Raddaoui Marín <notifications@github.com>
wrote:

> Thanks @sevein <https://github.com/sevein>,
>
> I've moved the tests and changed the pytest ignore rules. By adding the
> __init__ modules to the test packages inside each app, they are
> recognized by Django's test runner. See the current output
> <https://gist.github.com/jraddaoui/7cec90d023a5ea6e3d2010b4bae051b9>
> running manage.py test (sorry for the debug messages).
>
> As you suspected, some changes will be needed to run the tests in
> src/dashboard/tests with Django's runner. We could tweak the discovery to
> look for tests in src/dashboard instead of in src/dashboard/src (the
> default top-level where manage.py is located). Or move the tests folder
> from src/dashboard/tests to src/dashboard/src/tests so they are found by
> default.
>
> I tried the latter (+ adding the __init__ module in the tests folder) and
> the results are not bad at all (see output
> <https://gist.github.com/jraddaoui/608e910c29485de87269b8e5cab54d5f>,
> again with a lot of debug, sorry). Nevertheless, it shows 5 failures from
> 81 tests, when the pytest run shows no failures from 102 tests, so this
> will require a little more effort than just the discovery issue and maybe
> it deserves its own issue.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/artefactual/archivematica/pull/1448?email_source=notifications&email_token=AAOLCXDYWNOQIWCDI3BBZT3P64O77A5CNFSM4H6NWQXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZWPU2Y#issuecomment-510458475>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOLCXGTAZ3KLHFPWMEK3A3P64O77ANCNFSM4H6NWQXA>
> .
>


-- 
Ross Spencer
Software Developer, Artefactual Systems Inc.
www.artefactual.com

sevein reviewed (approved):
LGTM, thanks!

ross-spencer commented on src/dashboard/src/components/administration/tests/test_dip_upload_configs.py:
```diff
@@ -1,28 +1,24 @@
-
 THIS_DIR = os.path.dirname(os.path.abspath(__file__))
-FIXTURES_DIR = os.path.abspath(os.path.join(THIS_DIR, "../../../tests/fixtures"))
+FIXTURES_DIR = os.path.abspath(os.path.join(THIS_DIR, "../../../../tests/fixtures"))
```
Hi @jraddaoui this has been approved so feel free to ignore this suggested change, but should this `FIXTURES_DIR` path be set using something a little more idiomatic/platform independent? (Same for the processing configuration test  below)
```python
FIXTURES_DIR = os.path.join(
    os.path.dirname(
        os.path.dirname(
            os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
        )
    ),
    "tests",
    "fixtures",
)
```
**NB.** This is just one way known to me, there may be a better approach.

Though, the complexity of this leads me back to your original question about where the tests should sit. In line with the tests belonging to their own applications, should the fixtures for each of the components be independent of each other too? In that case we'd create a fixtures dir for each of the components?

jraddaoui replied:
Good point @ross-spencer!

Not a big deal, but that made me realize that we were building the fixture paths in all tests, when that can be done directly in the Django settings. Please, take a look when you can to the last commit ;)

sevein replied:
That's a nice improvement Radda.

jraddaoui committed:
Dashboard: improve fixtures configuration

- Specify FIXTURE_DIRS setting using system agnostic paths.
- Do not include fixtures from archivematicaCommon.
- Import local instead of base settings in test settings to also include
  the debug changes and the fixture directories.
- Trust FIXTURE_DIRS setting in tests, which only makes necessary to
  include the filenames in the `fixtures` attribute.

jraddaoui committed:
Dashboard: fix administration and fpr tests

- Avoid database connection on `fpr.forms` module import. The connection
  was created to get the choices of some fields. However those choices
  are overwritten later in the forms constructor.
- Split, rename and move tests modules to be picked by `pytest`.
- Add `__init__` module to test packages to be picked by Django's runner.
- Change pytest ignore rules in tox.ini.
